### PR TITLE
[cluster-test] Allow basic cluster-test against libra-swarm

### LINKED ISSUE #1219: [Feature Request] Allow to run cluster test locally with libra swarm
Running
`cargo run -p libra-swarm -- -n 4`
Will start local libra cluster of 4 nodes.
It would be good for dev purposes if we could also run cluster_test(at least in some capacity) on those nodes.
It would not be possible to integrate with experiments that require, for example, rebooting machines, but it can integrate with transaction generator(`--emit-tx` flag).

When libra swarm starts, it output two essential lines
```
Faucet account created in (loaded from) file <path to mint key file>
...
To run the Libra CLI client in a separate process and connect to the validator nodes you just spawned, use this command:
	cargo run --bin client -- -a localhost -p <Port>
```

Ideally we want libra swarm to additionally print something like

```
To run transaction generator run
cargo run --bin cluster-test -- --mint-file <path to mint key file> --peers <localhost:Port> --emit-tx 
```

This essentially requires modification in two places
(1) In libra swarm - to output this hint message with port and path to mint file taken from generated libra swarm config
(2) In cluster_test - support running it locally with libra swarm
   (a) Add `--mint-file` arg and read file name from it, instead of `"mint.key"` in `tx_emitter.rs`
   (b) Add `ac_port` field to `Instance` class to allow custom ac ports instead of default 8000 (see TxEmitter::make_client)
   (c) In `Cluster::discovery` detect if `--peers` arg has host:port pairs in it, use those host:port instead of aws discovery.

### PR DESCRIPTION
**Pull request by ankushagarwal**
# Summary

This diff will allow us to point cluster-test to a local libra-swarm

libra-swarm changes:
* Print cluster-test command when we start libra-swarm

cluster-test changes:
* Add "mint-file" flag to cluster-test binary
* Update struct Cluster: make prometheus_ip an Option<String> and add mint_file: String
* Add from_host_port method to struct Cluster
* Add port: u32 field to struct Instance
* During ClusterUtil::setup check if all peers are of format host:port, in that case don't use AWS

### Created a swarm

```
% cargo run -p libra-swarm -- -n 4
   Compiling libra-swarm v0.1.0 (/Users/kush/git/libra/libra-swarm)
    Finished dev [unoptimized + debuginfo] target(s) in 9.49s
     Running `target/debug/libra-swarm -n 4`
Faucet account created in (loaded from) file "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/42d246ecba8cdbf6126c8db7a0015181/temp_faucet_keys"
Base directory containing logs and configs: Temporary(TempPath { path_buf: "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/78126c6f0ae98e423ea76d9c62983c35", persist: false })
To run the Libra CLI client in a separate process and connect to the validator nodes you just spawned, use this command:
	cargo run --bin client -- -a localhost -p 60575 -s "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/78126c6f0ae98e423ea76d9c62983c35/0/consensus_peers.config.toml" -m "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/42d246ecba8cdbf6126c8db7a0015181/temp_faucet_keys"
To run transaction generator run:
	cargo run --bin cluster-test -- --mint-file "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/42d246ecba8cdbf6126c8db7a0015181/temp_faucet_keys" --peers localhost:60575  --emit-tx --workplace <workplace-name>
CTRL-C to exit.
```

### Ran cluster-test binary against this swarm
```
% cargo run --bin cluster-test -- --mint-file "/var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/42d246ecba8cdbf6126c8db7a0015181/temp_faucet_keys" --peers localhost:60575  --emit-tx --workplace kush                           (130)
    Finished dev [unoptimized + debuginfo] target(s) in 0.76s
     Running `target/debug/cluster-test --mint-file /var/folders/pl/fwpsbmh93mj444lfn6xpyf800000gn/T/42d246ecba8cdbf6126c8db7a0015181/temp_faucet_keys --peers 'localhost:60575' --emit-tx --workplace kush`
Oct 10 11:09:18.576 INFO Minting accounts
Oct 10 11:09:23.228 INFO Mint is done
```

Closes #1219

### CODE REVIEW

File: libra-swarm/src/main.rs [hunk 1]
```diff
@@ -98,6 +98,15 @@ fn main() {
         validator_set_file,
         faucet_key_file_path,
     );
+    println!("To run transaction generator run:");
+    println!(
+        // WIP(ankush): Do we need workplace for local cluster-test?
+        "\tcargo run --bin cluster-test -- --mint-file {:?} --peers localhost:{}  --emit-tx --workplace <workplace-name>",
```
Comment (ankushagarwal) [inline libra-swarm/src/main.rs]: Question: Do we need a `--workplace` parameter when we run cluster-test against a libra-swarm?

File: testsuite/cluster-test/src/cluster.rs [hunk 2]
```diff
@@ -9,11 +9,25 @@ use std::{thread, time::Duration};
...
+    mint_file: String,
 }
 
 impl Cluster {
-    pub fn discover(aws: &Aws) -> failure::Result<Self> {
+    pub fn from_host_port(peers: Vec<(String, u32)>, mint_file: &str) -> Self {
+        let instances: Vec<Instance> = peers
+            .into_iter()
+            // WIP(ankush): What should be short_hash here?
+            .map(|host_port| Instance::new("".to_string(), host_port.0, host_port.1))
```
Comment (ankushagarwal) [inline testsuite/cluster-test/src/cluster.rs]: Question: What should be short_hash here?

File: testsuite/cluster-test/src/main.rs [hunk 3]
```diff
@@ -251,17 +285,21 @@ impl ClusterUtil {
     }
 
     fn print_stat(&self, window: Duration) -> failure::Result<(f64, f64)> {
+        // WIP: What should be the behaviour here?
+        if self.prometheus.is_none() {
```
Comment (ankushagarwal) [inline testsuite/cluster-test/src/main.rs]: Question: What should be the behaviour here when there is no Prometheus instance?

**Comment by ankushagarwal** [comment]
cc @andll @dmitri-perelman

File: libra-swarm/src/main.rs (hunk 1 above)
Comment (andll) [reply libra-swarm/src/main.rs]: Nope, `--workplace` only makes sense in AWS setup (it is used for discovery)

Comment (andll) [reply libra-swarm/src/main.rs]: But perhaps we do want to put all the peers here, not just one in `--peers`

File: testsuite/cluster-test/src/cluster.rs [hunk 4]
```diff
@@ -86,7 +99,8 @@ impl Cluster {
             prometheus_ip.ok_or_else(|| format_err!("Prometheus was not found in workplace"))?;
         Ok(Self {
             instances,
-            prometheus_ip,
+            prometheus_ip: Some(prometheus_ip),
+            mint_file: String::from(mint_file),
```
Comment (andll) [inline testsuite/cluster-test/src/cluster.rs]: Most common pattern in our codebase is to use `.to_string()` to convert from &str to String

File: testsuite/cluster-test/src/cluster.rs [hunk 5]
```diff
@@ -99,17 +113,21 @@ impl Cluster {
...
+    }
+
     pub fn get_instance(&self, name: &str) -> Option<&Instance> {
         self.instances
             .iter()
             .find(|instance| instance.short_hash() == name)
     }
 
-    pub fn sub_cluster(&self, ids: Vec<String>) -> Cluster {
+    pub fn sub_cluster(&self, ids: Vec<String>, mint_file: &str) -> Cluster {
```
Comment (andll) [inline testsuite/cluster-test/src/cluster.rs]: I don't think we have use case for getting subcluster with different mint file, we can just copy value from parent cluster...

File: testsuite/cluster-test/src/instance.rs [hunk 6]
```diff
@@ -65,6 +70,10 @@ impl Instance {
     pub fn ip(&self) -> &String {
         &self.ip
     }
+
+    pub fn port(&self) -> u32 {
```
Comment (andll) [inline testsuite/cluster-test/src/instance.rs]: We should probably specify that this is `ac_port`, because there are many other ports for instance - ssh port, debug port, etc

File: testsuite/cluster-test/src/main.rs [hunk 7]
```diff
@@ -144,41 +146,73 @@ fn setup_log() {
 
 struct ClusterUtil {
     cluster: Cluster,
-    aws: Aws,
-    prometheus: Prometheus,
+    aws: Option<Aws>,
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: So for now we only want to support `--emit-tx` for libra swarm, everything else will require both Prometheus and Aws.
Instead of making those field optional in ClusterUtil, I think it would be cleaner to make another struct, like `BasicSwarmUtil` or something, that will have minimal number of dependencies (Cluster and Option<Prometheus>) and move emit_tx function there.

For everything else we can keep ClusterUtil as is and avoid unwrapping in all the methods of it that do require aws.

As you can see we similarly have `ClusterUtil` and `ClusterTestRunner`. `ClusterUtil` has subset of fields of `ClusterTestRunner` and it contains commands that do not require many things that `ClusterTestRunner`.

So basically we going to have
`ClusterTestRunner` - to run comprehensive tests, which contains things like HealthCheckRunner and Deployment.
`ClusterUtil` - to run simple commands on Aws cluster, prometheus is required (not an `Option<>`).
`BasicSwarmUtil` - to run commands on arbitrary bunch of nodes, does not depend on Aws at all, and have optional Prometheus. For now will have `--emit-tx` only.

File: testsuite/cluster-test/src/main.rs [hunk 8]
```diff
@@ -144,41 +146,73 @@ fn setup_log() {
...
+    prometheus: Option<Prometheus>,
 }
 
 struct ClusterTestRunner {
     logs: LogTail,
     cluster: Cluster,
     health_check_runner: HealthCheckRunner,
-    deployment_manager: DeploymentManager,
+    deployment_manager: Option<DeploymentManager>,
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Similarly, whatever methods ClusterTestRunner has, should require aws and deployment_manager, so we can keep this field as non-optional

File: testsuite/cluster-test/src/main.rs (hunk 3 above)
Comment (andll) [reply testsuite/cluster-test/src/main.rs]: We can just return error with `bail!`

File: testsuite/cluster-test/src/cluster.rs (hunk 2 above)
Comment (andll) [reply testsuite/cluster-test/src/cluster.rs]: So I was thinking that --peers would contain `short_hash:ac_port`
For libra swarm host is always localhost, so we can assume that

File: testsuite/cluster-test/src/main.rs [hunk 9]
```diff
@@ -144,41 +146,73 @@ fn setup_log() {
...
+        let mut parsed_peers = vec![];
+        for peer in &args.peers {
+            if let Ok(res) = parse_host_port(peer) {
+                parsed_peers.push(res);
+            } else {
+                is_host_port_format = false;
+                break;
+            }
+        }
+        if !parsed_peers.is_empty() && is_host_port_format {
```
Comment (huitseeker) [inline testsuite/cluster-test/src/main.rs]: How about [this](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect)?
```
let res: Result<Vec<_>> = args.peers.iter().map(|peer| parse_host_port(peer)).collect();

if let OK(parsed_peers) = res {
```
?

File: testsuite/cluster-test/src/main.rs (hunk 3 above)
Comment (huitseeker) [reply testsuite/cluster-test/src/main.rs]: Consider `let my_prometheus = self.prometheus?;` or `let my_prometheus = self.prometheus.map_err(format_err!("<what you would pass as bail arg>"));`

Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Cleaned up so that `self.prometheus` is not an Option<T> anymore.

File: testsuite/cluster-test/src/main.rs (hunk 9 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Much better! Updated.

File: testsuite/cluster-test/src/main.rs (hunk 7 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Created `BasicSwarmUtil` and created an `emit_tx` there. It is similar to ClusterUtil::emit_tx with the difference that we don't call `run_stat_loop` at the end of BasicSwarmUtil::emit_tx.

I have not added Prometheus to `BasicSwarmUtil` for now  since we don't have Prometheus running with libra-swarm. Once we do, we can add that field.

**Comment by ankushagarwal** [comment]
@huitseeker

> Consider `let my_prometheus = self.prometheus?;` or `let my_prometheus = self.prometheus.map_err(|err| format_err!("<what you would pass as bail arg>", err))?;`

I have updated `struct ClusterUtil` so that prometheus is no longer an Option<T>

File: testsuite/cluster-test/src/main.rs [hunk 10]
```diff
@@ -37,11 +37,17 @@ const HEALTH_POLL_INTERVAL: Duration = Duration::from_secs(5);
 #[derive(StructOpt, Debug)]
 #[structopt(group = ArgGroup::with_name("action").required(true))]
 struct Args {
-    #[structopt(short = "w", long)]
+    #[structopt(short = "w", long, default_value = "")]
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Could be better to make workplace to be `Option<String>` and have None as default
Also can add `conflicts_with="swarm"`

File: testsuite/cluster-test/src/main.rs [hunk 11]
```diff
@@ -158,16 +180,56 @@ struct ClusterTestRunner {
...
+            .map(|peer| parse_host_port(peer).unwrap())
+            .collect();
+        Self {
+            cluster: Cluster::from_host_port(parsed_peers, &args.mint_file),
+        }
+    }
+
+    pub fn emit_tx(self, accounts_per_client: usize, thread_params: EmitThreadParams) {
+        let mut emitter = TxEmitter::new(&self.cluster);
+        let _job = emitter.start_job(EmitJobRequest {
```
Comment (andll) [inline testsuite/cluster-test/src/main.rs]: Did you check if it works? I am not super familiar with rust threading nuances, do other threads continue to work after main thread terminated?

File: testsuite/cluster-test/src/main.rs (hunk 10 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Updated.

File: testsuite/cluster-test/src/main.rs (hunk 11 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: You're right. When the main thread exits, the process exits taking down other threads along with it.

I have added `thread::park();` which blocks the main thread.
